OSDOCS-3717 update etcd recommendations for scalability and performance

lahinson opened the pull request:
<!--- PR title format: [GH#<gh-issue-id>][BZ#<bz-issue-id>][OSDOCS-<jira-issue-id>]: <short-description-of-the-pr> --->

Version(s): 4.6+

Issue: https://issues.redhat.com/browse/OSDOCS-3717

Link to docs preview: http://file.rdu.redhat.com/lahinson/etcd-ssd-recs-3717/scalability_and_performance/recommended-host-practices.html#recommended-etcd-practices_recommended-host-practices

Additional information:
Updated the "Recommended etcd practices" section of the "Recommended host practices" topic in the Scalability and performance docs. Cleaned up some of the language per IBM Style and included a new line that tells customers to consider using an SLC SSD for the most durability and reliability.



<!--- Next steps after opening your PR:

* Ask for peer review from the OpenShift docs team:
  - For Red Hat associates: Ping @peer-review-squad requesting a review in the #forum-docs-review channel (CoreOS Slack workspace) and provide the following information:
    * A link to the PR.
    * The size of the PR that the GitHub bot assigns (ex: XS, S, M, L, XL).
    * If there is urgency or a deadline for the review.
  - For community authors: Request a review by tagging @openshift/team-documentation in a GitHub comment.

  Slack is the quickest and preferred way to request a review.

* IMPORTANT:
  - All documentation changes must be verified by a QE team associate before merging.
  - Squash to one commit before submitting your PR for peer review.

* For more information about verifying your content, see:
  https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#verification-of-your-content

* For more information about contributing to OpenShift documentation, see:
  https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/contributing.adoc

Additional resources

The OpenShift docs repo adheres to the following style guides:

- OpenShift documentation guidelines (OSDOCS)
  https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc
- Red Hat Supplementary Style Guide (SSG)
  https://redhat-documentation.github.io/supplementary-style-guide/
- Modular Documentation Reference Guide (Mod Docs)
  https://redhat-documentation.github.io/modular-docs/
- IBM Style Guide (ISG)
  https://www.ibm.com/docs/en/ibm-style

You can log in to the ISG by using your @redhat.com id and single sign-on (SSO) credentials. --->

lahinson commented:
@tjungblu - When you can, please take a look at the update to the "Recommended etcd practices" section (see the link to the docs preview in the first comment 👆). Thanks!

lahinson commented:
Hi @geliu2016 - When you can, please take a look at the update to the "Recommended etcd practices" section (see the link to the docs preview in the first comment 👆). This change is related to [OSDOCS-3717](https://issues.redhat.com/browse/OSDOCS-3717). Thanks!

geliu2016 commented:
/lgtm

tjungblu commented:
just fyi, it's not lost @lahinson, I have to debug an issue today and won't get there today. I'll do this tomorrow first thing in the morning.

lahinson commented:
Thanks, @geliu2016!

And no worries, @tjungblu!

lahinson commented:
@tjungblu Just a friendly reminder to review this PR when you can. Thanks!

tjungblu commented:
@lahinson I reviewed it above last week

kelbrown20 commented on modules/recommended-etcd-practices.adoc:
```diff
@@ -6,59 +6,42 @@
 
-Fio, a I/O benchmarking tool can be used to validate the hardware for etcd before or after
-creating the {product-title} cluster. Run fio and analyze the results:
+Because etcd writes data to disk and persists proposals on disk, its performance depends on disk performance. Slow disks and disk activity from other processes can cause long fsync latencies, which can cause etcd to miss heartbeats, not commit new proposals to the disk on time, and ultimately experience request timeouts and temporary leader loss. Be sure to run etcd on machines that are backed by ...
```
```suggestion
Because etcd writes data to disk and persists proposals on disk, its performance depends on disk performance. Slow disks and disk activity from other processes can cause long fsync latencies, which can cause etcd to miss heartbeats, not commit new proposals to the disk on time, and ultimately experience request timeouts and temporary leader loss. Ensure you run etcd on machines that are backed by SSD or NVMe disks with low latency and high throughput. Consider single-level cell (SLC) solid-state drives (SSDs), which provide 1 bit per memory cell, are durable and reliable, and are ideal for write-intensive workloads.
```

kelbrown20 replied:
Up to you!

kelbrown20 commented on modules/recommended-etcd-practices.adoc:
```diff
@@ -6,59 +6,42 @@
+
+* The `etcd_disk_wal_fsync_duration_seconds_bucket` metric reports the etcd disk fsync duration.
+* The `etcd_server_leader_changes_seen_total` metric reports the leader changes.
+* To rule out a slow disk and confirm that the disk is reasonably fast, make sure that the 99th percentile of the `etcd_disk_wal_fsync_duration_seconds_bucket` is less than 10 ms.
```
Possibly expand the `ms`, but up to you!

kelbrown20 reviewed (commented):
Just a few minor nits, please let me know if you have questions on anything. Other than that, this looks great Laura!

lahinson commented:
@kelbrown20 Uh oh. I clicked the *Commit suggestion* button on a couple of your comments and just now saw the note in the OpenShift Docs Manual that says not to click that button. I don't see anything in the docs manual about what to do if you click that button. Am I doomed?

lahinson commented:
Hi @kelbrown20 - Thanks so much for the review and the help with git yesterday! I incorporated your suggestions and squashed my latest commits. Let me know if you need anything else from me!

kelbrown20 commented:
Looks great! If you have all your SME/QE reviews, you can send it over to the peer review squad for merging (I do not have merge rights yet)

kalexand-rh commented on modules/recommended-etcd-practices.adoc:
```diff
@@ -6,59 +6,44 @@
 
 .Procedure
-Run the following if using podman:
+If you use Podman, run this command:
```
Single steps need to start with `*`, and I think it would be more effective to make these step options under a  common step.

```
* Run fio and analyze the results: (I am assuming that this is the step. If so, remove this from line 27).
** If you use Podman, run this command:
+
----
command
----

** If you use Docker, run this command:
+
----
command
----
```

kalexand-rh replied:
(You can use the [grandparent list syntax](https://docs.asciidoctor.org/asciidoc/latest/lists/continuation/#enclose-in-open-block) to attach the note to the first level of the step.)

lahinson commented:
@kalexand-rh So, to be candid, the only thing I was really supposed to do here was to add this sentence to the module: "Consider single-level cell (SLC) solid-state drives (SSDs), which provide 1 bit per memory cell, are durable and reliable, and are ideal for write-intensive workloads."

I did that, but then I noticed some other style stuff that needed to be cleaned up, so I made a few more changes. I didn't know if I might be overstepping boundaries, so I tried to not touch too much. 

I see what you mean about the placement of the procedure, though. It looks like an author wedged a procedure about validating the hardware for etcd into what would otherwise be a concept module.  Do you think I should pull those steps into a new procedure module about validating hardware for etcd by using fio?

kalexand-rh commented:
@lahinson, it's fine to make other improvements in files if you have the time to do it, particularly if you're addressing style or structure issues. Making changes to improve clarity, localization, and minimalism are also fine, if you have time.

If you'd just added that line, I wouldn't have mentioned the change to the step structure - it would have been clearly out of scope.

I think that even with your changes, splitting the module would be out of scope. (You'd need to make sure that the pair of modules wound up in all the same places as the original one, and doing that can dramatically increase the PR scope.)

lahinson commented:
@kalexand-rh Thanks for your thoughts! I incorporated your suggestions and squashed my commits. I appreciate your patience with me while I learn. It's also good to know about the formatting guideline for one-step procedures. That differs a bit from IBM Style, so I wouldn't have known if you didn't mention it 👍

lahinson commented:
@kalexand-rh I think we're all good now! Thanks again for all of  your help. (And you know that I don't mind any fussiness over step styling 😄 )

kalexand-rh commented:
/cherrypick enterprise-4.8

kalexand-rh commented:
/cherrypick enterprise-4.9

kalexand-rh commented:
/cherrypick enterprise-4.10

kalexand-rh commented:
/cherrypick enterprise-4.11
